csrc/scheduler/pointwise_utils.cpp

```diff
@@ -142,6 +142,137 @@ bool DomainMap::areAllInputIdsMappedTo(TensorView* input_tv, TensorView* tv)
+
+    std::vector<IterDomain*> source_ids;
+    // filtering all producer IDs with empty definition to get source iter
+    // domains
+    std::for_each(
+        all_producer_sets.vector().begin(),
+        all_producer_sets.vector().end(),
+        [&source_ids,
+         this](const std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>&
+                   producer_set_ptr) {
+          IterDomain* producer_id = producer_set_ptr->front();
+          if (ca_map_.uniqueExactDefinitions(producer_id).empty()) {
+            source_ids.push_back(producer_id);
+          }
+        });
+    return source_ids;
+  };
+
+  // this contains all source iter domain that's covered by reference_tv, so
+  // it's safe for target_tv to have them.
+  std::unordered_set<IterDomain*> covered_source_ids;
+  for (IterDomain* source_id_ref : get_source_iter_domains(reference_tv)) {
+    covered_source_ids.insert(source_id_ref);
+  }
+  // It's safe to have unmapped broadcast IterDomain. There're quite a few tests
+  // expecting pointwise scheduler to handle this pattern
+  for (IterDomain* id_out : target_tv->getLogicalDomain()) {
+    if (id_out->isBroadcast()) {
+      NVF_ERROR(
+          id_out->definition() == nullptr ||
+          id_out->definition()->isA<Resize>());
+
+      // Note that ideally we should also be able to handle merge/split on
+      // broadcast IDs, so we should really move this skip inside the loop below
+      // `get_source_iter_domains(target_tv)` and skip broadcast source IDs.
+      // currently we have the issue that split/merge does not preserve expanded
+      // broadcasts, see issue: https://github.com/NVIDIA/Fuser/issues/1126
+      covered_source_ids.insert(id_out);
+    }
+  }
+  // Note: there's certain cases where it's safe to have dangling IDs,
+  // e.g
+  //   T34  [i0, i1]
+  //   T185 [i0, b2, i1]     = broadcast(T34)
+  //   T192 [i0, b3(ex), i1] = expand(T185)
+  // It's safe to propagate T34 to T192, since b3(ex) is not involved in the
+  // propagation. But this isn't generally safe. If the above example is changed
+  // to e.g
+  //   T34  [i0, i1]
+  //   T185 [i0, b2, i1]     = broadcast(T34)
+  //   T186 [i0, i4, i1]     = ones({i0, i4, i1})
+  //   T193 [i0, i4, i1]     = add(T185, T186)
+  // It's unsafe to propagate from T34 to T193, see issue
+  // https://github.com/NVIDIA/Fuser/issues/3542
+
+  // Check all source iter domain involved in producing target_tv
+  for (IterDomain* source_id_out : get_source_iter_domains(target_tv)) {
+    // NOTE: we use concrete id instead. This allows us to link indirect
+    // broadcast. So in the example below: T2[i0, i1] = T0[i0, b0] + T1[i0, i1]
+    // T3[i0, i9] = pad(T0[i0, b0])
+    // We have i9 in T3
+    //     -> source ID b0
+    //     -> concrete map to i1
+    // So T3 is contained by T2. See test `PointwiseTest.DomainMapPad1`
+    auto concrete_source_id_out =
+        ca_map_.getConcreteMappedID(source_id_out, IdMappingMode::PERMISSIVE);
```

Comment:
This is the change I made in order to avoid the regression in the added test. `PointwiseTest.DomainMapPad1`

Reply:
As I mentioned [here](https://github.com/NVIDIA/Fuser/issues/3576#issuecomment-2540643470), I'm not sure if it's considered a regression. 

How is a fusion like this scheduled?

Reply:
Scheduling looks pretty naive. It's just merging all the IDs on logical that's mapped and split by 128/ then 1.
```
(gdb) p reference_tv->toString(0)
$1 = "T3_g_float[iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}, ithreadIdx.x33{128}] ca_pos( 2 ) produce_pos( 3 )"
(gdb) p fusion->printMath(1)
Inputs:
  T0_g_float[iS72{( ceilDiv(i1, 128) )}, iS73{1}, iS71{128}]
  T1_g_float[iS62{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS63{1}, iS61{128}]
Outputs:
  T3_g_float[iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}, ithreadIdx.x33{128}] ca_pos( 2 ) produce_pos( 3 )
  T4_g_float[iblockIdx.x82{( ceilDiv(( i1 * 9 ), 128) )}, iUS83{1}, ithreadIdx.x81{128}] produce_pos( 3 )

%kernel_math {
T6_l_float[iblockIdx.x55{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iUS56{1}, ithreadIdx.x54{128}] ca_pos( 2 )
   = Set( T1_g_float[iS62{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS63{1}, iS61{128}], cache_op=AllLevels )
T5_l_float[iblockIdx.x67{( ceilDiv(i1, 128) )}, iUS68{1}, ithreadIdx.x66{128}] ca_pos( 2 )
   = Set( T0_g_float[iS72{( ceilDiv(i1, 128) )}, iS73{1}, iS71{128}], cache_op=AllLevels )
T2_l_float[iblockIdx.x48{( ceilDiv(i1, 128) )}, iUS49{1}, ithreadIdx.x47{128}] ca_pos( 3 ) produce_pos( 2 )
   = broadcast( T5_l_float[iblockIdx.x67{( ceilDiv(i1, 128) )}, iUS68{1}, ithreadIdx.x66{128}] ca_pos( 2 ) )
T7_l_float[iblockIdx.x41{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS42{1}, ithreadIdx.x40{128}] ca_pos( 3 ) produce_pos( 3 )
   = T6_l_float[iblockIdx.x55{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iUS56{1}, ithreadIdx.x54{128}] ca_pos( 2 )
   + T2_l_float[iblockIdx.x48{( ceilDiv(i1, 128) )}, iUS49{1}, ithreadIdx.x47{128}] ca_pos( 3 ) produce_pos( 2 );
T3_g_float[iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}, ithreadIdx.x33{128}] ca_pos( 2 ) produce_pos( 3 )
   = Set( T7_l_float[iblockIdx.x41{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS42{1}, ithreadIdx.x40{128}] ca_pos( 3 ) produce_pos( 3 ), cache_op=Streaming )
T8_l_float[iblockIdx.x77{( ceilDiv(( i1 * 9 ), 128) )}, iUS78{1}, ithreadIdx.x76{128}] ca_pos( 3 )
   = pad( T0_g_float[iS72{( ceilDiv(i1, 128) )}, iS73{1}, iS71{128}], {4, 4, 0, 0} )
T4_g_float[iblockIdx.x82{( ceilDiv(( i1 * 9 ), 128) )}, iUS83{1}, ithreadIdx.x81{128}] produce_pos( 3 )
   = Set( T8_l_float[iblockIdx.x77{( ceilDiv(( i1 * 9 ), 128) )}, iUS78{1}, ithreadIdx.x76{128}] ca_pos( 3 ), cache_op=Streaming )
} // %kernel_math
```

```
(gdb) p fusion->printTransforms()
T1_g_float[iS62{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS63{1}, iS61{128}]
 logical domain : (iS2{i4}, iS3{i5}, iS4{i6}, bS5{1})
 contiguity: f f f n
  Merge: iS4{i6} and bS5{1} -> iS57{i6}
  Merge: iS3{i5} and iS57{i6} -> iS58{( i5 * i6 )}
  Merge: iS2{i4} and iS58{( i5 * i6 )} -> iS59{( i4 * ( i5 * i6 ) )}
  Split: iS59{( i4 * ( i5 * i6 ) )} by factor 128 -> iS60{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS61{128}
  Split: iS60{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )} by factor 1 -> iS62{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS63{1}
 loop domain : (iS62{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iS63{1}, iS61{128})
T6_l_float[iblockIdx.x55{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iUS56{1}, ithreadIdx.x54{128}] ca_pos( 2 )
 logical domain : (iS19{i4}, iS20{i5}, iS21{i6}, bS22{1})
 contiguity: t t t n
  Merge: iS21{i6} and bS22{1} -> iS50{i6}
  Merge: iS20{i5} and iS50{i6} -> iS51{( i5 * i6 )}
  Merge: iS19{i4} and iS51{( i5 * i6 )} -> iS52{( i4 * ( i5 * i6 ) )}
  Split: iS52{( i4 * ( i5 * i6 ) )} by factor 128 -> iS53{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, ithreadIdx.x54{128}
  Split: iS53{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )} by factor 1 -> iblockIdx.x55{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iUS56{1}
 loop domain : (iblockIdx.x55{( ceilDiv(( i4 * ( i5 * i6 ) ), 128) )}, iUS56{1}, ithreadIdx.x54{128})
T0_g_float[iS72{( ceilDiv(i1, 128) )}, iS73{1}, iS71{128}]
 logical domain : (bS0{1}, iS1{i1})
 contiguity: n f
  Merge: bS0{1} and iS1{i1} -> iS69{i1}
  Split: iS69{i1} by factor 128 -> iS70{( ceilDiv(i1, 128) )}, iS71{128}
  Split: iS70{( ceilDiv(i1, 128) )} by factor 1 -> iS72{( ceilDiv(i1, 128) )}, iS73{1}
 loop domain : (iS72{( ceilDiv(i1, 128) )}, iS73{1}, iS71{128})
T5_l_float[iblockIdx.x67{( ceilDiv(i1, 128) )}, iUS68{1}, ithreadIdx.x66{128}] ca_pos( 2 )
 logical domain : (bS17{1}, iS18{i1})
 contiguity: n t
  Merge: bS17{1} and iS18{i1} -> iS64{i1}
  Split: iS64{i1} by factor 128 -> iS65{( ceilDiv(i1, 128) )}, ithreadIdx.x66{128}
  Split: iS65{( ceilDiv(i1, 128) )} by factor 1 -> iblockIdx.x67{( ceilDiv(i1, 128) )}, iUS68{1}
 loop domain : (iblockIdx.x67{( ceilDiv(i1, 128) )}, iUS68{1}, ithreadIdx.x66{128})
T2_l_float[iblockIdx.x48{( ceilDiv(i1, 128) )}, iUS49{1}, ithreadIdx.x47{128}] ca_pos( 3 ) produce_pos( 2 )
 logical domain : (bS6{1}, bS7{1}, bS8{1}, iS9{i1})
 contiguity: n n n t
  Merge: bS8{1} and iS9{i1} -> iS43{i1}
  Merge: bS7{1} and iS43{i1} -> iS44{i1}
  Merge: bS6{1} and iS44{i1} -> iS45{i1}
  Split: iS45{i1} by factor 128 -> iS46{( ceilDiv(i1, 128) )}, ithreadIdx.x47{128}
  Split: iS46{( ceilDiv(i1, 128) )} by factor 1 -> iblockIdx.x48{( ceilDiv(i1, 128) )}, iUS49{1}
 loop domain : (iblockIdx.x48{( ceilDiv(i1, 128) )}, iUS49{1}, ithreadIdx.x47{128})
T7_l_float[iblockIdx.x41{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS42{1}, ithreadIdx.x40{128}] ca_pos( 3 ) produce_pos( 3 )
 logical domain : (iS10{i4}, iS11{i5}, iS12{i6}, iS13{i1})
 contiguity: t t t t
  Merge: iS12{i6} and iS13{i1} -> iS36{( i6 * i1 )}
  Merge: iS11{i5} and iS36{( i6 * i1 )} -> iS37{( i5 * ( i6 * i1 ) )}
  Merge: iS10{i4} and iS37{( i5 * ( i6 * i1 ) )} -> iS38{( i4 * ( i5 * ( i6 * i1 ) ) )}
  Split: iS38{( i4 * ( i5 * ( i6 * i1 ) ) )} by factor 128 -> iS39{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, ithreadIdx.x40{128}
  Split: iS39{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )} by factor 1 -> iblockIdx.x41{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS42{1}
 loop domain : (iblockIdx.x41{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS42{1}, ithreadIdx.x40{128})
T3_g_float[iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}, ithreadIdx.x33{128}] ca_pos( 2 ) produce_pos( 3 )
 logical domain : (iS23{i4}, iS24{i5}, iS25{i6}, iS26{i1})
 contiguity: t t t t
  Merge: iS25{i6} and iS26{i1} -> iS29{( i6 * i1 )}
  Merge: iS24{i5} and iS29{( i6 * i1 )} -> iS30{( i5 * ( i6 * i1 ) )}
  Merge: iS23{i4} and iS30{( i5 * ( i6 * i1 ) )} -> iS31{( i4 * ( i5 * ( i6 * i1 ) ) )}
  Split: iS31{( i4 * ( i5 * ( i6 * i1 ) ) )} by factor 128 -> iS32{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, ithreadIdx.x33{128}
  Split: iS32{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )} by factor 1 -> iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}
 loop domain : (iblockIdx.x34{( ceilDiv(( i4 * ( i5 * ( i6 * i1 ) ) ), 128) )}, iUS35{1}, ithreadIdx.x33{128})
T8_l_float[iblockIdx.x77{( ceilDiv(( i1 * 9 ), 128) )}, iUS78{1}, ithreadIdx.x76{128}] ca_pos( 3 )
 root domain : (bS14{1}rf, iS16{i1})
  Resize: bS14{1}rf by 4 and 4 -> iS15{9}rf
 logical domain : (iS15{9}rf, iS16{i1})
 contiguity: t t
  Merge: iS15{9}rf and iS16{i1} -> iS74{( i1 * 9 )}
  Split: iS74{( i1 * 9 )} by factor 128 -> iS75{( ceilDiv(( i1 * 9 ), 128) )}, ithreadIdx.x76{128}
  Split: iS75{( ceilDiv(( i1 * 9 ), 128) )} by factor 1 -> iblockIdx.x77{( ceilDiv(( i1 * 9 ), 128) )}, iUS78{1}
 loop domain : (iblockIdx.x77{( ceilDiv(( i1 * 9 ), 128) )}, iUS78{1}, ithreadIdx.x76{128})
T4_g_float[iblockIdx.x82{( ceilDiv(( i1 * 9 ), 128) )}, iUS83{1}, ithreadIdx.x81{128}] produce_pos( 3 )
 logical domain : (iS27{9}, iS28{i1})
 contiguity: t t
  Merge: iS27{9} and iS28{i1} -> iS79{( i1 * 9 )}
  Split: iS79{( i1 * 9 )} by factor 128 -> iS80{( ceilDiv(( i1 * 9 ), 128) )}, ithreadIdx.x81{128}
  Split: iS80{( ceilDiv(( i1 * 9 ), 128) )} by factor 1 -> iblockIdx.x82{( ceilDiv(( i1 * 9 ), 128) )}, iUS83{1}
 loop domain : (iblockIdx.x82{( ceilDiv(( i1 * 9 ), 128) )}, iUS83{1}, ithreadIdx.x81{128})
```

Reply:
Generated kernel looks like this:
```
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 2, 2> T0, Tensor<float, 4, 4> T1, Tensor<float, 4, 4> T3, Tensor<float, 2, 2> T4) {
  nvfuser_index_t i0;
  i0 = ((nvfuser_index_t)threadIdx.x) + (128LL * ((nvfuser_index_t)blockIdx.x));
  nvfuser_index_t i1;
  i1 = (T0.logical_size[1LL] * T1.logical_size[1LL]) * T1.logical_size[2LL];
  nvfuser_index_t i2;
  i2 = i0 % i1;
  nvfuser_index_t i3;
  i3 = T0.logical_size[1LL] * T1.logical_size[2LL];
  nvfuser_index_t i4;
  i4 = 9LL * T0.logical_size[1LL];
  nvfuser_index_t i5;
  i5 = ((T1.logical_size[0LL] * T1.logical_size[1LL]) * T1.logical_size[2LL]) * T0.logical_size[1LL];
  if (((((nvfuser_index_t)blockIdx.x) < (ceilDiv(i4, 128LL))) && (i0 < i4))) {
    float T8[1LL];
    T8[0LL]
       = (((i0 / T0.logical_size[1LL]) >= 4LL) && (i0 < (5LL * T0.logical_size[1LL]))) ? T0[(T0.alloc_stride[1LL] * (i0 % T0.logical_size[1LL]))] : 0.000000000e+00f;
    T4[i0]
       = T8[0LL];
  }
  if (((((nvfuser_index_t)blockIdx.x) < (ceilDiv(i5, 128LL))) && (i0 < i5))) {
    float T5[1LL];
    T5[0LL] = 0LL;
    T5[0LL]
       = T0[(T0.alloc_stride[1LL] * (((i0 % ((T1.logical_size[1LL] * T1.logical_size[2LL]) * T0.logical_size[1LL])) % (T1.logical_size[2LL] * T0.logical_size[1LL])) % T0.logical_size[1LL]))];
    float T6[1LL];
    T6[0LL] = 0LL;
    T6[0LL]
       = T1[(((T1.alloc_stride[0LL] * (i0 / i1)) + (T1.alloc_stride[1LL] * (i2 / i3))) + (T1.alloc_stride[2LL] * ((i2 % i3) / T0.logical_size[1LL])))];
    float T2[1LL];
    T2[0LL]
       = T5[0LL];
    float T7[1LL];
    T7[0LL]
      = T6[0LL]
      + T2[0LL];
    T3[i0]
       = T7[0LL];
  }
}
```